pkg/blobstore/quorum/quorum_blob_access.go

```diff
@@ -0,0 +1,218 @@
+    if len(eh.remainingBackends) > 0 {
+        nextBackend := eh.remainingBackends[0]
+        eh.remainingBackends = eh.remainingBackends[1:]
+        return eh.retry(nextBackend), nil
+    }
+    return nil, err
+}
+
+func (eh *getQuorumErrorHandler) OnError(err error) (buffer.Buffer, error) {
+    fallbackErr := status.Error(codes.Unavailable, "Too many backends unavailable")
+    if util.IsInfrastructureError(err) {
+        // I/O error.  Try again on another backend.
+        return eh.tryNextBackendOrError(fallbackErr)
+
+    } else if status.Code(err) == codes.NotFound {
+        // Not found.  Try again on another backend - if we haven't seen enough yet.
+        if eh.remainingReads <= 1 {
+            // Observed sufficient NotFounds.  Return conclusive NotFound.
+            return nil, err
+        }
+        eh.remainingReads--
+
+        // Haven't been able to check enough backends.  Can't conclude not found.
+        return eh.tryNextBackendOrError(fallbackErr)
+    }
+
+	return nil, err
+}
+
+func (eh getQuorumErrorHandler) Done() {}
+
+func (ba *quorumBlobAccess) get(getter func(b blobstore.BlobAccess) buffer.Buffer) buffer.Buffer {
+	backends := ba.shuffledBackends()
+
+	backend := backends[0]
+	remainingBackends := backends[1:]
+
+	return buffer.WithErrorHandler(
+		getter(backend),
+		&getQuorumErrorHandler{
+			remainingBackends: remainingBackends,
+			remainingReads:    ba.readQuorum,
+			retry:             getter,
+		})
+}
+
+func (ba *quorumBlobAccess) Get(ctx context.Context, digest digest.Digest) buffer.Buffer {
+	return ba.get(func(b blobstore.BlobAccess) buffer.Buffer {
+		return b.Get(ctx, digest)
+	})
+}
+
+func (ba *quorumBlobAccess) GetFromComposite(ctx context.Context, parentDigest, childDigest digest.Digest, slicer slicing.BlobSlicer) buffer.Buffer {
+	return ba.get(func(b blobstore.BlobAccess) buffer.Buffer {
+		return b.GetFromComposite(ctx, parentDigest, childDigest, slicer)
+	})
+}
+
+func (ba *quorumBlobAccess) shuffledBackendQueue() <-chan blobstore.BlobAccess {
+	queue := make(chan blobstore.BlobAccess)
+	go func() error {
+		backends := ba.shuffledBackends()
+
+		for _, b := range backends {
+			queue <- b
+		}
+
+		close(queue)
+		return nil
+	}()
+    return queue
+}
+
+func (ba *quorumBlobAccess) Put(ctx context.Context, digest digest.Digest, b buffer.Buffer) error {
+	// Store object in at least writeQuorum storage backends.
+	group, ctx := errgroup.WithContext(ctx)
+    backendQueue := ba.shuffledBackendQueue()
+
+	// Spawn writeQuorum writers.  Each of these goroutines needs to succeed once.
+	for i := 0; i < ba.writeQuorum; i++ {
+		var b1 buffer.Buffer
+		if i == ba.writeQuorum-1 {
+			// Last writer, no need to clone buffer
+			b1 = b
+		} else {
+			b, b1 = b.CloneStream()
+		}
+
+		group.Go(func() error {
+			var err error
+			for backend := range backendQueue {
+				err = backend.Put(ctx, digest, b1)
```

Comment:
BlobAccess.Put() takes ownership of `b1`. You can't call it again with the same buffer. The reason being that it could have converted the buffer to a reader and already consumed some of the data inside. We can't rewind, as we may be streaming data coming from the client/server directly.